editor/test/editor/defold_project_search_test.clj

```diff
@@ -8,69 +8,77 @@
             [integration.test-util :as test-util]
             [support.test-support :refer [with-clean-system]]
             [util.thread-util :as thread-util])
-  (:import [java.util.concurrent LinkedBlockingQueue]))
+  (:import [java.util.concurrent LinkedBlockingQueue TimeUnit]))
 
 (def ^:const search-project-path "test/resources/search_project")
-(def ^:const timeout-ms 15000)
+(def ^:const timeout-ms 1000)
 
 (defn- make-consumer [report-error!]
-  (atom {:consumed [] :future nil :report-error! report-error!}))
-
-(defn- consumer-start! [consumer-atom results-fn!]
-  (swap! consumer-atom
-         (fn [consumer]
-           (when (:future consumer)
-             (future-cancel (:future consumer)))
-           (-> consumer
-               (assoc :consumed [])
-               (assoc :future (future
-                                (try
-                                  (loop [last-response-time (System/nanoTime)]
-                                    (thread-util/throw-if-interrupted!)
-                                    (let [poll-time (System/nanoTime)
-                                          ret (loop [[result & more] (results-fn!)]
-                                                (if result
-                                                  (if (= ::project-search/done result)
-                                                    ::done
-                                                    (do (swap! consumer-atom update :consumed conj result)
-                                                        (recur more)))
-                                                  ::not-done))]
-                                      (when (= ret ::not-done)
-                                        (when (< (- poll-time last-response-time) (* 1000000 timeout-ms))
-                                          (Thread/sleep 10)
-                                          (recur last-response-time)))))
-                                  (catch InterruptedException _
-                                    nil)
-                                  (catch Throwable error
-                                    ((:report-error! consumer) error)
-                                    nil)))))))
-  consumer-atom)
-
-(defn- consumer-stop! [consumer-atom]
-  (swap! consumer-atom
-         (fn [consumer]
-           (when (:future consumer)
-             (future-cancel (:future consumer)))
-           consumer))
-  nil)
-
-(defn- consumer-started? [consumer-atom]
-  (-> consumer-atom deref :future some?))
-
-(defn- consumer-finished? [consumer-atom]
-  (if-let [pending-future (-> consumer-atom deref :future)]
+  {:consumed-atom (atom [])
+   :future-atom (atom nil)
```

Comment:
Broke up into two separate atoms to ensure there is no `swap!` retry contention. Now only the consumer thread invokes `swap!`.